test/api/test_config.py

```diff
@@ -0,0 +1,168 @@
+"""
+Copyright (c) 2020 Genome Research Limited
+
+Author: Aiden Neale <an12@sanger.ac.uk>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or (at
+your option) any later version.
+
+This program is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see https://www.gnu.org/licenses/
+"""
+
+import os
+import unittest
+from tempfile import TemporaryDirectory
+from core import typing as T, time
+from core.config import exception
+from api.config import Config, _validate, _schema
+
+
+class TestLoader(unittest.TestCase):
+    _tmp:TemporaryDirectory
+    _path:T.Path
+
+    def setUp(self) -> None:
+        self._tmp = TemporaryDirectory()
+        self._path = path = T.Path(self._tmp.name)
+
+        tmp_file = path / "config"
+        tmp_file.touch()
+
+
+    def tearDown(self) -> None:
+        self._tmp.cleanup()
+        del self._path
+
+
+
+    def test_branch_contents(self) -> None:
+        config = Config(T.Path("eg/.vaultrc"))
+        self.assertEqual(config.identity.groups.attributes.owners, "owner")
+        self.assertEqual(config.identity.users.dn, "ou=users,dc=example,dc=com")
+
+        self.assertEqual(config.persistence.postgres.port, 5432)
+        self.assertEqual(config.persistence.password, "abc123")
+
+        self.assertEqual(config.email.smtp.host, "mail.example.com")
+        self.assertEqual(config.deletion.warnings, [time.delta(days=10), time.delta(days=3), time.delta(days=1)])
+
+        self.assertEqual(config.archive.handler, T.Path("/path/to/executable"))
+
+
+
+    def test_build(self) -> None:
+        _path = self._path / "config"
+
+        self.assertTrue(Config(T.Path("eg/.vaultrc")))
+
+        _not_a_config = """
+This is an example
+"""
+        _bad_YAML = """
+identity:
+            ldap:
+        host: baz
+    port: quux
+"""
+
+        with open(_path, 'w') as file:
+            file.write(_not_a_config)
+
+        self.assertRaises(exception.InvalidConfiguration, Config._build, _path)
+
+        with open(_path, 'w') as file:
+            file.write(_bad_YAML)
+
+        self.assertRaises(exception.InvalidConfiguration, Config._build, _path)
+
+
+
+    def test_validation(self) -> None:
```

Comment:
You never check the failing states for validation. This is complicated by the fact that `Config`'s constructor automatically calls `_is_valid` when building from YAML and, if it fails, raises the same exception as you'd see if there was a problem with the YAML passing (making it difficult to disambiguate). You can get around this in the following ways:

* While the same exception class is raised, the message is different. You can use `unittest.TestCase.assertRaisesRegex` to check the message. Personally, I don't think this is a brilliant idea, because it deeply ties your tests to the implementation, making them brittle (i.e., benign changes to the implementation -- like changing the exception message -- would cause the test to fail without being rewritten).

* `Config._is_valid` is simply calling the `_validate` function, therefore you could test for validation failures by testing the `_validate` function directly with the appropriate input. This is less brittle than the previous option, but it still ties you to an internal (the prefixed underscore implies no guarantees on the stability of the interface) implementation detail.

* The ultimate solution would be to make `_validate` and the machinery around it a first-class citizen. However, given that this machinery is only used once, when parsing the configuration, that seems overkill. One has to be pragmatic in these things.

I would therefore suggest a slight modification on the first option:

* Create a new exception, in `core.config`, that is a subclass of `InvalidConfiguration`, solely for the purpose of disambiguating semantic errors (i.e., validity against a give schema) from syntactic ones.
* In the constructor for `Config`, when `_is_valid` is checked to raise the exception, if necessary, change it from `InvalidConfiguration` to your new exception.
* Now, in the test, you can simply check against the new exception, when given invalid (but syntactically correct) input, writing appropriate inputs to cover all failing states of the validator.